src/function/built_in_functions.cpp

```diff
@@ -487,6 +525,31 @@ void BuiltInFunctions::validateNonEmptyCandidateFunctions(
     }
 }
 
+void BuiltInFunctions::validateSpecialCases(std::vector<Function*>& candidateFunctions,
+    const std::string& name, const std::vector<LogicalType*>& inputTypes) {
+    // special case for add func
+    if (name == ADD_FUNC_NAME) {
+        auto targetType0 = candidateFunctions[0]->parameterTypeIDs[0];
+        auto targetType1 = candidateFunctions[0]->parameterTypeIDs[1];
+        auto inputType0 = inputTypes[0]->getLogicalTypeID();
+        auto inputType1 = inputTypes[1]->getLogicalTypeID();
+        if ((inputType0 != LogicalTypeID::STRING || inputType1 != LogicalTypeID::STRING) &&
+            targetType0 == LogicalTypeID::STRING && targetType1 == LogicalTypeID::STRING) {
+            if (inputType0 != inputType1 && (inputType0 == LogicalTypeID::RDF_VARIANT ||
+                                                inputType1 == LogicalTypeID::RDF_VARIANT)) {
+                return;
+            }
+            std::string supportedInputsString;
+            for (auto& function : functions.at(name)) {
+                supportedInputsString += function->signatureToString() + "\n";
+            }
+            throw BinderException("Cannot match a built-in function for given function " + name +
+                                  LogicalTypeUtils::toString(inputTypes) +
+                                  ". Supported inputs are\n" + supportedInputsString);
+        }
+    }
+}
+
```

Comment:
I retain string +string and use `validateSpecialCases`  for solving special cases like date+date and rdf_varient+string. I notice that we can not remove `hasimplict` function because of some special cases. Maybe this function can be used for those special cases later.